Refactor operator dispatch framework across different Tensors.

pritamdamania87 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #77707

Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

pritamdamania87 committed:
Refactor operator dispatch framework across different Tensors.

Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Refactor operator dispatch framework across different Tensors."


Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Refactor operator dispatch framework across different Tensors."


Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

[ghstack-poisoned]

pritamdamania87 committed:
Update on "Refactor operator dispatch framework across different Tensors."


Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

[ghstack-poisoned]

wanchaol commented on torch/distributed/_shard/partial_tensor.py:
```diff
@@ -1,27 +1,24 @@
+from torch.distributed._shard.op_registry_utils import _decorator_func
+from torch.utils._pytree import tree_map
+
+if TYPE_CHECKING:
```
it looks like we already do the conditional imports in the `reshard` function, maybe we don't need this import?

pritamdamania87 replied:
The "ShardedTensor" return annotation for `reshard` needs this logic.

wanchaol reviewed (approved):
looks good! have one minor comment

fduwjj reviewed (approved):
lgtm

pritamdamania87 commented:
@pytorchbot merge this please

pritamdamania87 committed:
Update on "Refactor operator dispatch framework across different Tensors."


Refactor to clean up the following pieces:

1) Consolidate decorators to use a common way to look up operator tables.
2) Move a bunch of utilities to `op_registry_utils` and `common_op_utils` and
reuse them across ShardedTensor, ReplicatedTensor and PartialTensor.

Differential Revision: [D36465639](https://our.internmc.facebook.com/intern/diff/D36465639/)

[ghstack-poisoned]

pritamdamania87 commented:
@pytorchbot merge this please
